refactor!: back nanoevents.methods.vector with scikit-hep vector

lgray opened the pull request:
Fixes #992

Right now this is breaking in terms of some low-level definitions of physics variables, in particular:
- `rho` is now *always* transverse, as opposed to it being previously `r`, this can certainly screw up math if you're not expecting it

Breaking interface differences (as you can see in tests):
- we're now beholden to vector's much more strict typing of vectors you have to use `to_VectorND()` to correctly compare vectors.

But that doesn't seem to have too many side effects?

However, all the tests pass and the current implementation is backwards compatible in terms of `delta_phi` vs `deltaphi` and `delta_r` vs. `deltaR` so we can actually deprecate that in a reasonable way.

Another thing we need to address is where the functions `nearest` and `metric_table` should live, which are fairly often used but could be advertised better.

@nsmith- This will require a thorough review as well as contribution from you, and we should discuss how we want to roll this out.

lgray committed:
first try at backing nanoevents.methods.vector with scikit-hep vector

lgray committed:
first try at backing nanoevents.methods.vector with scikit-hep vector

nsmith- commented on src/coffea/nanoevents/methods/vector.py:
```diff
@@ -669,53 +641,14 @@ def pvec(self):
     def boostvec(self):
         """The `x`, `y` and `z` components divided by `t` as a `ThreeVector`
 
         This can be used for boosting. For cases where `|t| <= rho`, this
```
This is now `r` in the docstring, right?

lgray replied:
ah yeah.

lgray replied:
Actually, also need to make sure that the sign in the metric is the same for coffea vs. scikit-hep vector.

nsmith- commented on tests/test_nanoevents_vector.py:
```diff
@@ -444,12 +444,18 @@ def test_lorentz_vector_numba(a_dtype, b_dtype):
-        0.6078019961139605,
-    ]
+
+    computed_dphi = a.delta_phi(b).to_numpy()
```
Is the explicit cast required now?

lgray replied:
see the comment on approx.

nsmith- commented on tests/test_nanoevents_vector.py:
```diff
@@ -444,12 +444,18 @@ def test_lorentz_vector_numba(a_dtype, b_dtype):
+
+    computed_dphi = a.delta_phi(b).to_numpy()
+
+    assert pytest.approx(computed_dphi, abs=1e-6) == np.array(
```
Any idea why the tolerance had to be increased?

lgray replied:
It needs further investigation but yeah it looks like there's some conversion going on or something that we should investigate. 

This is also related to the above, I was just making sure there wasn't weird roundoff error happening or anything.

nsmith- commented on tests/test_nanoevents_vector.py:
```diff
@@ -532,9 +538,9 @@ def test_inherited_method_transpose(lcoord, threecoord, twocoord):
-    assert record_arrays_equal(b - c, -(c - b))
+    assert record_arrays_equal((a - b).to_Vector3D(), -(b - a))
+    assert record_arrays_equal((a - c).to_Vector2D(), -(c - a))
+    assert record_arrays_equal((b - c).to_Vector2D(), -(c - b))
```
`a` is a 4-vector, `b` a 3-vector, and `c` a 2-vector. So I guess the type demotion behavior depends on only the first argument of the infix operator here? Is this expected?

lgray replied:
@saransh-cpp do we expect this (3D to 2D vector demotion / promotion)?

Saransh-cpp replied:
Yes, the behavior is uniform across all backends.

lgray replied:
In that you expect when you do `==` between two vectors you have to cast them to the same dimensionality first?


@nsmith- what was the expected behavior in nanoevents.methods.vector?

Saransh-cpp replied:
Yes, for `==` the vectors should have same dimensions. `to_Vector*D` or `to_{coordinate_names}` methods can be used to change the dimensions. `to_Vector*D` additionally allows users to update the newly created coordinate (when the vector's dimension is promoted).

nsmith- replied:
So `-(c-a)` being automatically demoted to a 2-vector makes sense, but I'm puzzled why `(a-c)` is _not_ and requires an explicit cast.

lgray replied:
My opinion is you should promote to the highest vector in the operation and then cast down in dimensionality as you need. That has the most reasonable behavior from my perspective.

However right now we do neither!

nsmith- replied:
I disagree. By promoting the 2-vector to a 3- or 4-vector, you're assuming a value of 0 for the missing dimension(s). If someone wants to add MET to a 4-vector they should have to explicitly upcast it from a 2-vector and define explicitly the pseudorapidity and mass.
I also don't think that `a - c` and `c - a` should result in different types.

lgray replied:
`a - c` vs. `c - a` implies some sort of promotion or demotion has to happen. So you are asking that we always demote?

Then common patterns like `some_4_vector + 2_vector_spatial_rotation` no longer result in a 4-vector, which would be very confusing.

lgray replied:
Oh, I misread. You're asking that in order to add a rotation someone does `4_vector + 2_vector.to_Vector4D()` always and specifically, or something even different from that? This is a breaking change we should make sure people know it is coming for the shim-in.

However, if we don't require that explicit dimensionality cast, that means we must choose which direction we go, since `a-c` vs. `c-a` have different starting types.
Do we always choose 0s for the lower dim vector?

nsmith- replied:
Yes I think we should always demote. I don't understand what you mean by adding a rotation. A rotation depends on the current vector and some (euler) angles. Anyway the current coffea vectors demote always:
```
>>> a-c
<TwoVectorArray [{x: 20, y: -22}, ..., {x: 15, ...}] type='3 * TwoVector[x:...'>
>>> c-a
<TwoVectorArray [{x: -20, y: 22}, ..., {x: -15, ...}] type='3 * TwoVector[x...'>
>>> a+c
<TwoVectorArray [{x: 0, y: 2}, {...}, {x: 45, ...}] type='3 * TwoVector[x: ...'>
>>> c+a
<TwoVectorArray [{x: 0, y: 2}, {...}, {x: 45, ...}] type='3 * TwoVector[x: ...'>
>>> b-c
<TwoVectorArray [{x: 0, y: -22}, ..., {x: -45, ...}] type='3 * TwoVector[x:...'>
>>> b+c
<TwoVectorArray [{x: -20, y: 2}, ..., {x: -15, ...}] type='3 * TwoVector[x:...'>
>>> c-b
<TwoVectorArray [{x: 0, y: 22}, ..., {x: 45, ...}] type='3 * TwoVector[x: f...'>
>>> c+b
<TwoVectorArray [{x: -20, y: 2}, ..., {x: -15, ...}] type='3 * TwoVector[x:...'>
>>> a+b
<ThreeVectorArray [{x: 0, y: -20, z: 10}, ..., {...}] type='3 * ThreeVector...'>
```

lgray replied:
Arg, morning brain glitch(re: rotations).

nsmith- replied:
@Saransh-cpp I think this is just a bug in vector: https://github.com/scikit-hep/vector/issues/412

lgray replied:
There's also an override in the nanoevents classes, that right now has weird behavior, we can just enforce it correctly there. That's what I was planning to do.

nsmith- commented on src/coffea/nanoevents/methods/vector.py:
```diff
@@ -91,10 +97,39 @@ def delta_r(eta1, phi1, eta2, phi2):
+behavior.update(vector.backends.awkward.behavior)
+
+
+class TwoVectorArray:
```
What do these empty classes do?

lgray replied:
These satisfy the linter and are overwritten by `@awkward.mixin_class(behavior)`, otherwise flake complains the variable doesn't exist since it's patched in and I have to assign the projection classes.

lgray replied:
changed this to use noqa instead, it's more readable.

nsmith- reviewed (changes requested):
I guess the only painful part is `r` vs. `rho`. Otherwise looks like it's drag and drop. @jrueb might want to review this as well.

lgray committed:
fix math in doc

lgray committed:
fix math in doc

lgray committed:
use noqa instead of actually satisfying flake - more readable

lgray committed:
use noqa instead of actually satisfying flake - more readable

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

nsmith- committed:
Improve tests of vector

nsmith- committed:
Improve tests of vector

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

lgray committed:
appease linter

lgray committed:
appease linter

nsmith- committed:
Manually build binary ufunc dispatch tables

Eventually we might also want to pawn these off on vector

nsmith- committed:
Manually build binary ufunc dispatch tables

Eventually we might also want to pawn these off on vector

lgray committed:
skooch pin for vector

lgray committed:
skooch pin for vector

Saransh-cpp commented on src/coffea/nanoevents/methods/vector.py:
```diff
@@ -1061,6 +912,52 @@ def negative(self):
+
+PtEtaPhiELorentzVectorArray.ProjectionClass2D = TwoVectorArray  # noqa: F821
+PtEtaPhiELorentzVectorArray.ProjectionClass3D = ThreeVectorArray  # noqa: F821
+PtEtaPhiELorentzVectorArray.ProjectionClass4D = LorentzVectorArray  # noqa: F821
```
Should this be? -
```suggestion
TwoVectorArray.ProjectionClass2D = TwoVectorArray  # noqa: F821
TwoVectorArray.ProjectionClass3D = ThreeVectorArray  # noqa: F821
TwoVectorArray.ProjectionClass4D = LorentzVectorArray  # noqa: F821

PolarTwoVectorArray.ProjectionClass2D = TwoVectorArray  # noqa: F821
PolarTwoVectorArray.ProjectionClass3D = ThreeVectorArray  # noqa: F821
PolarTwoVectorArray.ProjectionClass4D = LorentzVectorArray  # noqa: F821

ThreeVectorArray.ProjectionClass2D = TwoVectorArray  # noqa: F821
ThreeVectorArray.ProjectionClass3D = ThreeVectorArray  # noqa: F821
ThreeVectorArray.ProjectionClass4D = LorentzVectorArray  # noqa: F821

SphericalThreeVectorArray.ProjectionClass2D = TwoVectorArray  # noqa: F821
SphericalThreeVectorArray.ProjectionClass3D = ThreeVectorArray  # noqa: F821
SphericalThreeVectorArray.ProjectionClass4D = LorentzVectorArray  # noqa: F821

LorentzVectorArray.ProjectionClass2D = TwoVectorArray  # noqa: F821
LorentzVectorArray.ProjectionClass3D = ThreeVectorArray  # noqa: F821
LorentzVectorArray.ProjectionClass4D = LorentzVectorArray  # noqa: F821

PtEtaPhiMLorentzVectorArray.ProjectionClass2D = TwoVectorArray  # noqa: F821
PtEtaPhiMLorentzVectorArray.ProjectionClass3D = ThreeVectorArray  # noqa: F821
PtEtaPhiMLorentzVectorArray.ProjectionClass4D = LorentzVectorArray  # noqa: F821

PtEtaPhiELorentzVectorArray.ProjectionClass2D = TwoVectorArray  # noqa: F821
PtEtaPhiELorentzVectorArray.ProjectionClass3D = ThreeVectorArray  # noqa: F821
PtEtaPhiELorentzVectorArray.ProjectionClass4D = LorentzVectorArray  # noqa: F821
```

nsmith- replied:
So obviously whatever solution we pick should make sure the array type matches the available coordinates.
That said, there is a choice here, whether:
- `SphericalThreeVector + ThreeVector = ThreeVector`, or
- `SphericalThreeVector + ThreeVector = SphericalThreeVector`

and similarly whether transposing the arguments maintains this casting.

Personally, and I've mentioned this to @jpivarski in the past, I would have had everything convert to cartesian on addition, even as far as to have `SphericalThreeVector + SphericalThreeVector = ThreeVector` (as coffea does now). The motivation is that adding is simplest in cartesian coordinates and I fear that doing it in other coordinates may inflate the effects of numerical imprecision in some cases. This would need to be carefully studied of course.

For now, I would advocate for add/subtract to return cartesian coordinates.

jpivarski replied:
In Vector, longitudinal and temporal coordinates each convert to Cartesian (`z` and `t`, respectively); it's only the polar azimuthal coordinates that stay polar if both addends are already polar. That's [this function](https://github.com/scikit-hep/vector/blob/e4c3f536d8d3833ba8b31453dc7313b5b3cc5f56/src/vector/_compute/planar/add.py#L46-L54):

```python
# specialized
def rhophi_rhophi(lib, rho1, phi1, rho2, phi2):
    diff = phi2 - phi1
    u = rho2 * lib.cos(diff)
    v = rho2 * lib.sin(diff)
    return (
        lib.sqrt((rho1 + u) ** 2 + v**2),
        rectify(lib, phi1 + lib.arctan2(v, rho1 + u)),
    )
```

If this is bad (no study was performed—what metric would define "goodness"?[^1]), then it can be replaced by

```python
def rhophi_rhophi(lib, rho1, phi1, rho2, phi2):
    return xy_xy(
        lib,
        x.rhophi(lib, rho1, phi1),
        y.rhophi(lib, rho1, phi1),
        x.rhophi(lib, rho2, phi2),
        y.rhophi(lib, rho2, phi2),
    )
```

and similarly for [subtract](https://github.com/scikit-hep/vector/blob/e4c3f536d8d3833ba8b31453dc7313b5b3cc5f56/src/vector/_compute/planar/subtract.py#L46-L54).

[^1]: Arguably, it could just be the count of transcendental functions. The `x.rhophi` and `y.rhophi` functions each have one `sin` and `cos`, so that's four, and the specialized function has a `cos`, `sin`, `sqrt`, and `arctan2`, but also the `rectify`, which uses `%` to keep the phi from wrapping around. I bet the conversion to Cartesian is faster for all inputs. To quantify round-off error, it might depend on what the inputs are (a 4D space of two pairs of azimuthal coordinates).

Saransh-cpp replied:
Thanks @nsmith-  and @jpivarski! I'd be happy to change the way things are handled in vector or keep them as it is, but in any case, I think we should get rid off `add`, `subtract`, and other methods in coffea's vector classes. Please let me know if vector should always convert azimuthal coordinates to cartesian.

nsmith- replied:
Can you try an experiment? Sample randomly two Lorentz vectors in cartesian space (set `t^2=x^2+y^2+z^2+m^2` for some randomly sampled m), then compare the result of: adding them; converting them to azimuthal coordinates and adding them? The sampling distribution could just be gaussian with a large (~50) width.

jpivarski replied:
Accuracy and/or speed?

For accuracy, note that among Python, NumPy, and Awkward Array, there are either 2 or 3 summation algorithms that differ in the least significant bits, so choice of backend matters.

For speed, naturally big arrays are more representative.

**Edit:** I'm hooked. I'll give it a try.

Since this is a question about how only the azimuthal coordinates are handled, I'm going to do the test in x and y only. That will focus on the part that might or might not be changed.

jpivarski replied:
Accuracy with NumPy:

```python
>>> import numpy as np
>>> import awkward as ak
>>> import vector
>>> vector.register_awkward()
>>> 
>>> source_ax = np.random.normal(0, 1, 1000000)
>>> source_ay = np.random.normal(0, 1, 1000000)
>>> source_bx = np.random.normal(0, 1, 1000000)
>>> source_by = np.random.normal(0, 1, 1000000)
>>> 
>>> cartesian_a = vector.array({"x": source_ax, "y": source_ay})
>>> cartesian_b = vector.array({"x": source_bx, "y": source_by})
>>> cartesian_sum = cartesian_a + cartesian_b
>>> 
>>> polar_a = vector.array({"rho": cartesian_a.rho, "phi": cartesian_a.phi})
>>> polar_b = vector.array({"rho": cartesian_b.rho, "phi": cartesian_b.phi})
>>> polar_sum = polar_a + polar_b
>>> 
>>> abs(cartesian_sum - polar_sum).min()
0.0
>>> abs(cartesian_sum - polar_sum).mean()
4.489121610871359e-16
>>> abs(cartesian_sum - polar_sum).max()
5.748513452393584e-15
```

Accuracy with Awkward:

```python
>>> cartesian_a = ak.zip({"x": source_ax, "y": source_ay}, with_name="Vector2D")
>>> cartesian_b = ak.zip({"x": source_bx, "y": source_by}, with_name="Vector2D")
>>> cartesian_sum = cartesian_a + cartesian_b
>>> 
>>> polar_a = ak.zip({"rho": cartesian_a.rho, "phi": cartesian_a.phi}, with_name="Vector2D")
>>> polar_b = ak.zip({"rho": cartesian_b.rho, "phi": cartesian_b.phi}, with_name="Vector2D")
>>> polar_sum = polar_a + polar_b
>>> 
>>> ak.min(abs(cartesian_sum - polar_sum))
0.0
>>> ak.mean(abs(cartesian_sum - polar_sum))
4.489131490391888e-16
>>> ak.max(abs(cartesian_sum - polar_sum))
5.748513452393584e-15
```

Accuracy with Python:

```python
>>> cartesian_a = [vector.obj(x=x, y=y) for x, y in zip(source_ax, source_ay)]
>>> cartesian_b = [vector.obj(x=x, y=y) for x, y in zip(source_bx, source_by)]
>>> cartesian_sum = [a + b for a, b in zip(cartesian_a, cartesian_b)]
>>> 
>>> polar_a = [vector.obj(rho=a.rho, phi=a.phi) for a in cartesian_a]
>>> polar_b = [vector.obj(rho=b.rho, phi=b.phi) for b in cartesian_b]
>>> polar_sum = [a + b for a, b in zip(polar_a, polar_b)]
>>> 
>>> tmp = [abs(car - pol) for car, pol in zip(cartesian_sum, polar_sum)]
>>> np.min(tmp)
0.0
>>> np.mean(tmp)
4.489131490397885e-16
>>> np.max(tmp)
5.748513452393584e-15
```

Although there were a million elementwise `a`, `b` pairs, these accuracy measurements represent 1 numerical step. The error would grow if this is compounded in a loop. (But when are you going to have coordinates changing between Cartesian and polar in a loop? Once some operation puts it in Cartesian, it will remain in Cartesian; that's an absorbing state.)

**Edit:** do-over with the same random numbers in all three studies, to see if any algorithms differ. It seems that they don't! (Okay, Awkward and NumPy only differ in repeated summations, as in `ak.sum`, scikit-hep/awkward#1250. It's not a surprise that NumPy and Python agree, since `vector.obj` functions [are computed with](https://github.com/scikit-hep/vector/blob/6ee3a1f81f02ce8013c6a766860ed05009d672fe/src/vector/backends/object.py#L336) `lib=numpy`, rather than `math` or something.)

jpivarski replied:
As for time, performing the calculation in polar is only 2% slower than converting to Cartesian, then adding.

```python
>>> import numpy as np
>>> import vector
>>> import time
>>> source_ax = np.random.normal(0, 1, 100000000)
>>> source_ay = np.random.normal(0, 1, 100000000)
>>> source_bx = np.random.normal(0, 1, 100000000)
>>> source_by = np.random.normal(0, 1, 100000000)
>>> 
>>> cartesian_a = vector.array({"x": source_ax, "y": source_ay})
>>> cartesian_b = vector.array({"x": source_bx, "y": source_by})
>>> polar_a = vector.array({"rho": cartesian_a.rho, "phi": cartesian_a.phi})
>>> polar_b = vector.array({"rho": cartesian_b.rho, "phi": cartesian_b.phi})
>>> 
>>> t0 = time.perf_counter(); tmp = polar_a + polar_b; time.perf_counter() - t0
6.290486244004569
>>> t0 = time.perf_counter(); tmp = polar_a + polar_b; time.perf_counter() - t0
6.331587403001322
>>> t0 = time.perf_counter(); tmp = polar_a + polar_b; time.perf_counter() - t0
6.302776351003558
>>> 
>>> t0 = time.perf_counter(); tmp = vector.array({"x": polar_a.x, "y": polar_a.y}) + vector.array({"x": polar_b.x, "y": polar_b.y}); time.perf_counter() - t0
6.182807716999378
>>> t0 = time.perf_counter(); tmp = vector.array({"x": polar_a.x, "y": polar_a.y}) + vector.array({"x": polar_b.x, "y": polar_b.y}); time.perf_counter() - t0
6.175768633998814
>>> t0 = time.perf_counter(); tmp = vector.array({"x": polar_a.x, "y": polar_a.y}) + vector.array({"x": polar_b.x, "y": polar_b.y}); time.perf_counter() - t0
6.201813601001049
```

By comparison, if it had been Cartesian to begin with, that's 8.8× faster:

```python
>>> t0 = time.perf_counter(); tmp = cartesian_a + cartesian_b; time.perf_counter() - t0
0.7100841389983543
>>> t0 = time.perf_counter(); tmp = cartesian_a + cartesian_b; time.perf_counter() - t0
0.710351111003547
>>> t0 = time.perf_counter(); tmp = cartesian_a + cartesian_b; time.perf_counter() - t0
0.6980755539989332
```

An argument could be made to convert to Cartesian before adding, not to speed up the addition, but to speed up any _subsequent_ operations. It has long-range performance repercussions, not short-range.

But leaving polar coordinates in polar was motivated by the principle of least surprise. An expert user might know that converting to Cartesian is better for their particular computation in the long run, but it's best to leave that decision to them.

jpivarski replied:
These are all of the constituent parts, the cost of each operation (ignoring memory access):

```python
>>> numbers = abs(source_ax)
>>> t0 = time.perf_counter(); tmp = numbers + numbers; time.perf_counter() - t0
0.11195918399607763
>>> t0 = time.perf_counter(); tmp = numbers * numbers; time.perf_counter() - t0
0.11245346799842082
>>> t0 = time.perf_counter(); tmp = numbers**2; time.perf_counter() - t0
0.11356563799927244
>>> t0 = time.perf_counter(); tmp = np.sin(numbers); time.perf_counter() - t0
0.8857823760044994
>>> t0 = time.perf_counter(); tmp = np.cos(numbers); time.perf_counter() - t0
0.9174221430002945
>>> t0 = time.perf_counter(); tmp = np.sqrt(numbers); time.perf_counter() - t0
0.16433826299908105
>>> t0 = time.perf_counter(); tmp = np.arctan2(numbers, numbers); time.perf_counter() - t0
1.1808518600009847
>>> def rectify(lib, phi):
...     return (phi + lib.pi) % (2 * lib.pi) - lib.pi
... 
>>> t0 = time.perf_counter(); tmp = rectify(np, numbers); time.perf_counter() - t0
0.5352489489960135
```

The additions and multiplications are likely dominated by memory access, so 0.11 seconds is the floor. `np.arctan2` is a large part of the 6.3 seconds.

nsmith- replied:
Ok, thanks for checking. I'm ok then with leaving it to the user to decide when to change coordinates. This does mean that the current coffea vector behavior of `PtEtaPhiMLorentzVector + PtEtaPhiMLorentzVector = LorentzVector`will change. We are opting in to this now.

Saransh-cpp reviewed (commented):
I tried removing the `add` and `subtract` methods locally and noticed something peculiar (notice the types and the coordinate names) -

## This branch
```py
In [3]: a = ak.zip(
   ...:             {
   ...:                 "pt": [10.0, 20.0, 30.0],
   ...:                 "eta": [0.0, 1.1, 2.2],
   ...:                 "phi": [0.1, 0.9, -1.1],
   ...:                 "mass": [1.0, 1.0, 1.0],
   ...:             },
   ...:             with_name="PtEtaPhiMLorentzVector",
   ...:             behavior=vector.behavior,
   ...:         )

In [4]: b = ak.zip(
   ...:             {
   ...:                 "rho": [10.0, 20.0, 30.0],
   ...:                 "theta": [0.3, 0.6, 1.1],
   ...:                 "phi": [-3.0, 1.1, 0.2],
   ...:             },
   ...:             with_name="SphericalThreeVector",
   ...:             behavior=vector.behavior,
   ...:         )

In [5]: c = ak.zip(
   ...:             {"rho": [-10.0, 13.0, 15.0], "phi": [1.22, -1.0, 1.0]},
   ...:             with_name="PolarTwoVector",
   ...:             behavior=vector.behavior,
   ...:         )

In [6]: a + b
Out[6]: <ThreeVectorArray [{x: 0.0501, y: -0.413, ...}, ...] type='3 * ThreeVector[...'>

In [7]: a + c
Out[7]: <TwoVectorArray [{x: 6.51, y: -8.39}, {...}, {...}] type='3 * TwoVector[x: ...'>

In [8]: c + a
Out[8]: <TwoVectorArray [{x: 6.51, y: -8.39}, {...}, {...}] type='3 * TwoVector[x: ...'>

In [9]: b + a
Out[9]: <ThreeVectorArray [{x: 0.0501, y: -0.413, ...}, ...] type='3 * ThreeVector[...'>
```

## Removing the `add` and `subtract` methods
```py
In [3]: a = ak.zip(
   ...:             {
   ...:                 "pt": [10.0, 20.0, 30.0],
   ...:                 "eta": [0.0, 1.1, 2.2],
   ...:                 "phi": [0.1, 0.9, -1.1],
   ...:                 "mass": [1.0, 1.0, 1.0],
   ...:             },
   ...:             with_name="PtEtaPhiMLorentzVector",
   ...:             behavior=vector.behavior,
   ...:         )

In [4]: b = ak.zip(
   ...:             {
   ...:                 "rho": [10.0, 20.0, 30.0],
   ...:                 "theta": [0.3, 0.6, 1.1],
   ...:                 "phi": [-3.0, 1.1, 0.2],
   ...:             },
   ...:             with_name="SphericalThreeVector",
   ...:             behavior=vector.behavior,
   ...:         )

In [5]: c = ak.zip(
   ...:             {"rho": [-10.0, 13.0, 15.0], "phi": [1.22, -1.0, 1.0]},
   ...:             with_name="PolarTwoVector",
   ...:             behavior=vector.behavior,
   ...:         )

In [6]: a + b
Out[6]: <ThreeVectorArray [{x: 0.0501, y: -0.413, ...}, ...] type='3 * ThreeVector[...'>

In [7]: b + a
Out[7]: <SphericalThreeVectorArray [{x: 0.0501, y: -0.413, ...}, ...] type='3 * Sph...'>

In [8]: a + c
Out[8]: <TwoVectorArray [{rho: 10.6, phi: -0.911}, ..., {...}] type='3 * TwoVector[...'>

In [9]: c + a
Out[9]: <PolarTwoVectorArray [{rho: 10.6, phi: -0.911}, ..., {...}] type='3 * Polar...'>
```

## Removing `add` and `subtract` and incorporating the suggestion below
```py
In [3]: a = ak.zip(
   ...:             {
   ...:                 "pt": [10.0, 20.0, 30.0],
   ...:                 "eta": [0.0, 1.1, 2.2],
   ...:                 "phi": [0.1, 0.9, -1.1],
   ...:                 "mass": [1.0, 1.0, 1.0],
   ...:             },
   ...:             with_name="PtEtaPhiMLorentzVector",
   ...:             behavior=vector.behavior,
   ...:         )

In [4]: b = ak.zip(
   ...:             {
   ...:                 "rho": [10.0, 20.0, 30.0],
   ...:                 "theta": [0.3, 0.6, 1.1],
   ...:                 "phi": [-3.0, 1.1, 0.2],
   ...:             },
   ...:             with_name="SphericalThreeVector",
   ...:             behavior=vector.behavior,
   ...:         )

In [5]: c = ak.zip(
   ...:             {"rho": [-10.0, 13.0, 15.0], "phi": [1.22, -1.0, 1.0]},
   ...:             with_name="PolarTwoVector",
   ...:             behavior=vector.behavior,
   ...:         )

In [6]: a = b

In [7]: a + b
Out[7]: <ThreeVectorArray [{rho: 20, phi: -3, ...}, ..., {...}] type='3 * ThreeVect...'>

In [8]: b + a
Out[8]: <ThreeVectorArray [{rho: 20, phi: -3, ...}, ..., {...}] type='3 * ThreeVect...'>

In [9]: a + c
Out[9]: <TwoVectorArray [{rho: 17.2, phi: -2.46}, ..., {...}] type='3 * TwoVector[r...'>

In [10]: c + a
Out[10]: <TwoVectorArray [{rho: 17.2, phi: -2.46}, ..., {...}] type='3 * TwoVector[r...'>
```

I'm not sure which one of these is correct, could you please let me know which one is? All of this is most probably on coffea's side and not on vector's side.

lgray committed:
skooch pin for vector

lgray committed:
skooch pin for vector

lgray committed:
Merge branch 'master' into use_scikithep_vector

Saransh-cpp committed:
Depend on vector v1.3 + cleanup

Saransh-cpp committed:
Depend on vector v1.3 + cleanup

Saransh-cpp committed:
Candidate classes need MmomentumClass too

Saransh-cpp committed:
Candidate classes need MmomentumClass too

Saransh-cpp committed:
Depend on vector v1.3.1

Saransh-cpp committed:
Depend on vector v1.3.1

Saransh-cpp committed:
MomentumClass for Jet and Candidate

Saransh-cpp committed:
MomentumClass for Jet and Candidate

lgray committed:
Merge pull request #1061 from Saransh-cpp/use_scikithep_vector

fix: depend on vector v1.3.1 + trim down coffea vectors

lgray committed:
Merge branch 'master' into use_scikithep_vector

Saransh-cpp reviewed (approved):
This looks good to me in this state!

lgray committed:
Merge branch 'master' into use_scikithep_vector

Saransh-cpp committed:
fix: add correct behaviors and projection classes for candidate classes

Saransh-cpp committed:
typo

Saransh-cpp committed:
fix behavior

lgray committed:
Merge branch 'master' into use_scikithep_vector

Saransh-cpp committed:
use copy_behaviors to fix ufuncs

Saransh-cpp committed:
Merge branch 'use_scikithep_vector' into fix-candidate-behaviors

Saransh-cpp committed:
Typo

lgray committed:
Merge pull request #1063 from Saransh-cpp/fix-candidate-behaviors

fix: add correct behaviors and projection classes for candidate classes

lgray committed:
Merge branch 'master' into use_scikithep_vector

Saransh-cpp committed:
fix: copy behaviors for the remaining vector sub-classes

lgray committed:
Merge pull request #1124 from Saransh-cpp/more-behavior-copying

fix: copy behaviors for the remaining vector sub-classes

lgray committed:
awkward 2.6.6 is required

jpivarski reviewed (approved):
This is such a clean swap: impressive!

lgray committed:
Merge branch 'master' into use_scikithep_vector

lgray committed:
update LowPtElectron

lgray committed:
Merge branch 'master' into use_scikithep_vector

lgray committed:
Merge branch 'master' into use_scikithep_vector

lgray committed:
Merge branch 'master' into use_scikithep_vector

nsmith- commented on src/coffea/nanoevents/methods/candidate.py:
```diff
@@ -70,4 +70,23 @@ class PtEtaPhiECandidate(Candidate, vector.PtEtaPhiELorentzVector):
     pass
 
 
+awkward.behavior.update(
```
Why are we updating the global awkward behavior and not our module-local `behavior` object defined on L13?

nsmith- commented on src/coffea/nanoevents/methods/delphes.py:
```diff
@@ -130,6 +130,14 @@ def z(self):
 
 
 _set_repr_name("Vertex")
+awkward.behavior.update(
```
Same comment as before, we are registering in the global namespace when there already is a local namespace that's wired into the NanoEvents

nsmith- commented on src/coffea/nanoevents/methods/delphes.py:
```diff
@@ -130,6 +130,14 @@ def z(self):
+
+VertexArray.ProjectionClass2D = vector.TwoVectorArray  # noqa: F821
+VertexArray.ProjectionClass3D = vector.ThreeVectorArray  # noqa: F821
+VertexArray.ProjectionClass4D = VertexArray  # noqa: F821
```
Is this a way to signal that there is no 4D interpretation of VertexArray? (which is true)

Saransh-cpp replied:
I can just remove L139, but vector would throw an error saying something like "ProjectionClass4D is not defined for VertexArray" if a user tries an operation that needs converting the array to a 4D vector. The error does not tell the user what they are doing wrong, but I am hoping that not a lot of users will try converting it to 4D?

nsmith- replied:
I think it might make sense to raise an error if a user tries to write a code that requires implicit upcast from 3D point to 4D

Saransh-cpp replied:
Thanks! I will open an issue in vector for this.

nsmith- commented on src/coffea/nanoevents/methods/candidate.py:
```diff
@@ -70,4 +70,23 @@ class PtEtaPhiECandidate(Candidate, vector.PtEtaPhiELorentzVector):
+    awkward._util.copy_behaviors(vector.LorentzVector, Candidate, behavior)
+)
+
+CandidateArray.ProjectionClass2D = vector.TwoVectorArray  # noqa: F821
+CandidateArray.ProjectionClass3D = vector.ThreeVectorArray  # noqa: F821
+CandidateArray.ProjectionClass4D = vector.LorentzVectorArray  # noqa: F821
+CandidateArray.MomentumClass = CandidateArray  # noqa: F821
```
This block of items is commonly repeated throughout. It would be good to create a utility function to set these projections

nsmith- commented on tests/test_nanoevents_vector.py:
```diff
@@ -519,22 +534,22 @@ def test_inherited_method_transpose(lcoord, threecoord, twocoord):
+    assert_record_arrays_equal((a.like(b) - b), -(b - a.like(b)), check_type=True)
+    assert_record_arrays_equal((a.like(c) - c), -(c - a.like(c)), check_type=True)
+    assert_record_arrays_equal((b.like(c) - c), -(c - b.like(c)), check_type=True)
 
```
I would suggest also testing we error on incompatible types:
```python
with pytest.raises(ValueError):
    a - b == -(b - a)
```

nsmith- reviewed (changes requested):
There are a few cases where we are adding to the `ak.behavior` global namespace. In general we try to keep everything in our own namespace. For example, it would break the awkward namespace if someone had a Delphes and NanoAOD file open at the same time since both schemas end up defining a `Jet` object mixin that has a different implementation. I am wondering how tests do not pick up on this, as the copy_behaviors output would not end up in the schema's behavior and I don't think the `ak.behavior` is queried at all for arrays built using NanoEventsFactory.

There are a few more minor comments related to code duplication, and I'm mildly concerned how the tests are all approximate now. In how many cases did we actually lose exact agreement?

Saransh-cpp committed:
Review changes

- a new util function to register projection classes
- do not modify the global awkward behavior
- tests for geometric dimension errors

Saransh-cpp committed:
Merge branch 'use_scikithep_vector' of https://github.com/CoffeaTeam/coffea into use_scikithep_vector

lgray committed:
Merge pull request #1130 from Saransh-cpp/use_scikithep_vector

fix: updates from review

lgray committed:
Merge branch 'master' into use_scikithep_vector

Saransh-cpp committed:
Revert "fix: updates from review"

Saransh-cpp committed:
Add back useful changes

* More tests

* Don't modify the global awkward behavior

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Saransh-cpp committed:
Indentation

lgray committed:
Merge pull request #1133 from Saransh-cpp/revert-1130-use_scikithep_vector

fix: Revert "fix: updates from review"

Saransh-cpp committed:
fix: use the new type signature for copy_behaviors

Saransh-cpp committed:
Update src/coffea/nanoevents/methods/nanoaod.py

lgray committed:
Merge pull request #1136 from Saransh-cpp/new-copy_behaviors

fix: use the new type signature for copy_behaviors

lgray committed:
Merge branch 'master' into use_scikithep_vector

lgray committed:
Merge branch 'master' into use_scikithep_vector

lgray committed:
use python -m pytest (windows path change!?)

lgray committed:
Merge branch 'master' into use_scikithep_vector

Saransh-cpp committed:
fix: copy_behaviors should be called before class definition

lgray committed:
Merge pull request #1152 from Saransh-cpp/fix-tests

fix: copy_behaviors should be called before class definition

lgray committed:
bump date of complete vector deprecation

nsmith- commented on tests/test_nanoevents_vector.py:
```diff
@@ -519,22 +534,36 @@ def test_inherited_method_transpose(lcoord, threecoord, twocoord):
     elif twocoord == "PolarTwoVector":
         c = ak.zip(
-            {"r": [-10.0, 13.0, 15.0], "phi": [1.22, -1.0, 1.0]},
+            {"rho": [-10.0, 13.0, 15.0], "phi": [1.22, -1.0, 1.0]},
```
This renaming of `r` to `rho` is something we'll need to remember to include in the release notes.
Perhaps a migration guide is in order? Are there any other user-facing changes we'll have to advertise?

lgray replied:
This the only one that isn't resolved by reading vector's documentation.

I think this one will be a fairly low hit-rate, since most people would specify a polar two vector using pT and phi, which is the same in both cases.

nsmith- reviewed (approved):
Looks good!
